pkg/rtc/dynacastmanager.go

```diff
@@ -0,0 +1,265 @@
+	d.maxSubscribedQuality[mime] = maxQuality
+	d.lock.Unlock()
+
+	d.update(false)
+}
+
+func (d *DynacastManager) update(force bool) {
+	d.lock.Lock()
+
+	changed := false
+	downgradesOnly := true
+	for mime, quality := range d.maxSubscribedQuality {
+		if cq, ok := d.committedMaxSubscribedQuality[mime]; !ok {
+			// a new mime has been added
+			changed = true
+			downgradesOnly = false
+		} else {
+			if cq != quality {
+				changed = true
+			}
+
+			if (cq == livekit.VideoQuality_OFF && quality != livekit.VideoQuality_OFF) || (cq != livekit.VideoQuality_OFF && quality != livekit.VideoQuality_OFF && cq < quality) {
+				downgradesOnly = false
+			}
+		}
+	}
+
+	if !force {
+		if !changed {
+			d.lock.Unlock()
+			return
+		}
+
+		if downgradesOnly {
+			d.params.Logger.Debugw("debouncing quality downgrade",
+				"committedMaxSubscribedQuality", d.committedMaxSubscribedQuality,
+				"maxSubscribedQuality", d.maxSubscribedQuality,
+			)
+			d.maxSubscribedQualityDebounce(func() {
+				d.update(true)
+			})
+			d.lock.Unlock()
+			return
+		}
+	}
+
+	// clear debounce on send
+	d.maxSubscribedQualityDebounce(func() {})
+
+	d.params.Logger.Infow("committing quality change",
+		"force", force,
+		"committedMaxSubscribedQuality", d.committedMaxSubscribedQuality,
+		"maxSubscribedQuality", d.maxSubscribedQuality,
+	)
+
+	// commit change
+	d.committedMaxSubscribedQuality = make(map[string]livekit.VideoQuality, len(d.maxSubscribedQuality))
+	for mime, quality := range d.maxSubscribedQuality {
+		d.committedMaxSubscribedQuality[mime] = quality
+	}
+
+	d.enqueueSubscribedQualityChange()
+	d.lock.Unlock()
+}
+
+func (d *DynacastManager) enqueueSubscribedQualityChange() {
+	if d.isClosed {
+		return
+	}
+
+	subscribedCodec := make([]*livekit.SubscribedCodec, 0, len(d.committedMaxSubscribedQuality))
+	maxSubscribedQualities := make([]types.SubscribedCodecQuality, 0, len(d.committedMaxSubscribedQuality))
+	for mime, quality := range d.committedMaxSubscribedQuality {
+		maxSubscribedQualities = append(maxSubscribedQualities, types.SubscribedCodecQuality{
+			CodecMime: mime,
+			Quality:   quality,
+		})
+
+		if quality == livekit.VideoQuality_OFF {
+			subscribedCodec = append(subscribedCodec, &livekit.SubscribedCodec{
+				Codec: mime,
+				Qualities: []*livekit.SubscribedQuality{
+					{Quality: livekit.VideoQuality_LOW, Enabled: false},
+					{Quality: livekit.VideoQuality_MEDIUM, Enabled: false},
+					{Quality: livekit.VideoQuality_HIGH, Enabled: false},
+				},
+			})
+		} else {
+			var subscribedQualities []*livekit.SubscribedQuality
+			for q := livekit.VideoQuality_LOW; q <= livekit.VideoQuality_HIGH; q++ {
+				subscribedQualities = append(subscribedQualities, &livekit.SubscribedQuality{
+					Quality: q,
+					Enabled: q <= quality,
+				})
+			}
+			subscribedCodec = append(subscribedCodec, &livekit.SubscribedCodec{
+				Codec:     mime,
+				Qualities: subscribedQualities,
+			})
+		}
+	}
+	if d.onSubscribedMaxQualityChange != nil {
```

Comment:
can do this check before calculate subscribedCodecs

Reply:
Will change.